src/hdmf/testing/testcase.py

```diff
@@ -62,49 +64,68 @@ def assertContainerEqual(self, container1, container2, ignore_name=False, ignore
         elif isinstance(f1, dict) and len(f1) and isinstance(f1.values()[0], Container):
             self.assertIsInstance(f2, dict)
             f1_keys = set(f1.keys())
             f2_keys = set(f2.keys())
             self.assertSetEqual(f1_keys, f2_keys)
             for k in f1_keys:
                 with self.subTest(module_name=k):
-                    self.assertContainerEqual(f1[k], f2[k], ignore_hdmf_attrs=ignore_hdmf_attrs)
+                    self.assertContainerEqual(f1[k], f2[k],
+                                              ignore_hdmf_attrs=ignore_hdmf_attrs,
+                                              ignore_string_to_byte=ignore_string_to_byte)
         elif isinstance(f1, Container):
-            self.assertContainerEqual(f1, f2, ignore_hdmf_attrs=ignore_hdmf_attrs)
+            self.assertContainerEqual(f1, f2,
+                                      ignore_hdmf_attrs=ignore_hdmf_attrs,
+                                      ignore_string_to_byte=ignore_string_to_byte)
         elif isinstance(f1, Data):
-            self._assert_data_equal(f1, f2, ignore_hdmf_attrs=ignore_hdmf_attrs)
+            self._assert_data_equal(f1, f2,
+                                    ignore_hdmf_attrs=ignore_hdmf_attrs,
+                                    ignore_string_to_byte=ignore_string_to_byte)
         elif isinstance(f1, (float, np.floating)):
             np.testing.assert_allclose(f1, f2)
         else:
             self.assertEqual(f1, f2)
 
-    def _assert_data_equal(self, data1, data2, ignore_hdmf_attrs=False):
+    def _assert_data_equal(self, data1, data2, ignore_hdmf_attrs=False, ignore_string_to_byte=False):
         self.assertTrue(isinstance(data1, Data))
         self.assertTrue(isinstance(data2, Data))
         self.assertEqual(len(data1), len(data2))
-        self._assert_array_equal(data1.data, data2.data, ignore_hdmf_attrs=ignore_hdmf_attrs)
+        self._assert_array_equal(data1.data, data2.data,
+                                 ignore_hdmf_attrs=ignore_hdmf_attrs,
+                                 ignore_string_to_byte=ignore_string_to_byte)
         self.assertContainerEqual(data1, data2, ignore_hdmf_attrs=ignore_hdmf_attrs)
 
-    def _assert_array_equal(self, arr1, arr2, ignore_hdmf_attrs=False):
-        if isinstance(arr1, (h5py.Dataset, HDMFDataset)):
+    def _assert_array_equal(self, arr1, arr2, ignore_hdmf_attrs=False, ignore_string_to_byte=False):
+        array_data_types = tuple([i for i in get_docval_macros('array_data')
+                                  if (i != list and i != tuple and i != AbstractDataChunkIterator)])
+        # We construct array_data_types this way to avoid explicit depencendy on h5py, Zarr and other
+        # I/O backends. Only list and tuple do not support [()] slicing, and AbstractDataChunkIterator
+        # should never occur here. The effective value of array_data_types is then:
+        # array_data_types = (np.ndarry, h5py.Dataset, zarr.core.Array, HDMFDataset)
```

Comment:
```suggestion
        # We construct array_data_types this way to avoid explicit dependency on h5py, Zarr and other
        # I/O backends. Only list and tuple do not support [()] slicing, and AbstractDataChunkIterator
        # should never occur here. The effective value of array_data_types is then:
        # array_data_types = (np.ndarray, h5py.Dataset, zarr.core.Array, hdmf.query.HDMFDataset)
```